# **Pandas Introduction**
# What is Pandas?

Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

\

# Why Use Pandas?
Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

\

# What Can Pandas Do?
Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?
What is average value?
Max value?
Min value?
Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

\

**Where is the Pandas Codebase?**

The source code for Pandas is located at this github repository https://github.com/pandas-dev/pandas

**Import Pandas**

Once Pandas is installed, import it in your applications by adding the import keyword:

In [1]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


**Pandas as pd**

Pandas is usually imported under the pd alias.

alias: In Python alias are an alternate name for referring to the same thing.

Create an alias with the as keyword while importing:

In [2]:
import pandas as pd

# Now the Pandas package can be referred to as pd instead of pandas.


import pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


**Pandas Series**

**What is a Series?**

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [3]:
# Create a simple Pandas Series from a list:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

0    1
1    7
2    2
dtype: int64


**Labels**

If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [4]:
# Return the first value of the Series:

print(myvar[0])

1


**Create Labels**

With the index argument, you can name your own labels.

In [5]:
# Create your own labels:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


When you have created labels, you can access an item by referring to the label.


In [6]:
# Return the value of "y":

print(myvar["y"])

7


In [7]:
# Create your own labels for this 2d array:

import pandas as pd

arr = [[1, 2, 3], [4, 5, 6]]

# Code Here
myVar = pd.Series(arr, index=["a", "b"])
print(myVar)

a    [1, 2, 3]
b    [4, 5, 6]
dtype: object


In [8]:
# Return the value 5:

# Code Here
myVar["b"][1]

5

**Key/Value Objects as Series**

You can also use a key/value object, like a dictionary, when creating a Series.

In [9]:
# Create a simple Pandas Series from a dictionary:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)


day1    420
day2    380
day3    390
dtype: int64


Note: The keys of the dictionary become the labels.

To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

In [10]:
# Create a Series using only data from "day1" and "day2":

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


In [11]:
# Create a Series using only data from "b":

import pandas as pd

calories = {"a": 12, "b": 31, "c": 24}

# Code Here
myVar = pd.Series(calories, index=["b"])
print(myVar)

b    31
dtype: int64


**DataFrames**

Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

Series is like a column, a DataFrame is the whole table.

In [12]:
# Create a DataFrame from two Series:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


**What is a DataFrame?**

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [13]:
# Create a simple Pandas DataFrame:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 

   calories  duration
0       420        50
1       380        40
2       390        45


**Locate Row**

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)

In [14]:
# Return row 0:

#refer to the row index:
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [15]:
# Return row 0 and 1:

#use a list of indexes:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


In [16]:
# Return row 1 and 2:

#use a list of indexes:

# Code Here
print(df.loc[[1, 2]])

   calories  duration
1       380        40
2       390        45


**Named Indexes**

With the index argument, you can name your own indexes.

In [17]:
# Add a list of names to give each row a name:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df) 

      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [18]:
# Add a list of names to give each row a name:

import pandas as pd

data1 = {
  "a": [3, 6, 9, 12],
  "b": [2, 4, 6, 8],
  "c": [10, 20, 30, 50]
}

# Code Here

df1 = pd.DataFrame(data1, index=["Row1", "Row2", "Row3", "Row4"])
print(df1)

       a  b   c
Row1   3  2  10
Row2   6  4  20
Row3   9  6  30
Row4  12  8  50


**Locate Named Indexes**

Use the named index in the loc attribute to return the specified row(s).

In [19]:
# Return "day2":

# refer to the named index:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


In [20]:
# Return a row from df1:

# refer to the named index:

# Code Here
print(df1.loc["Row3"])

a     9
b     6
c    30
Name: Row3, dtype: int64


**Pandas Read CSV**

**Read CSV Files**

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be using a CSV file called 'data.csv'.

In [21]:
# Load the CSV into a DataFrame:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string()) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [22]:
# Load the "forestfires.csv" into a DataFrame:

# Code Here
import pandas as pd

forestfires = pd.read_csv("forestfires.csv")

use to_string() to print the entire DataFrame.

If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:


In [23]:
# Print the data DataFrame without the to_string() method:

import pandas as pd

print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


In [24]:
# Print the "forestfires" DataFrame without the to_string() method:

# Code Here
print(forestfires)

     X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain   area
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51.0   6.7   0.0   0.00
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33.0   0.9   0.0   0.00
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33.0   1.3   0.0   0.00
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97.0   4.0   0.2   0.00
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99.0   1.8   0.0   0.00
..  .. ..   ...  ...   ...    ...    ...   ...   ...   ...   ...   ...    ...
512  4  3   aug  sun  81.6   56.7  665.6   1.9  27.8  32.0   2.7   0.0   6.44
513  2  4   aug  sun  81.6   56.7  665.6   1.9  21.9  71.0   5.8   0.0  54.29
514  7  4   aug  sun  81.6   56.7  665.6   1.9  21.2  70.0   6.7   0.0  11.16
515  1  4   aug  sat  94.4  146.0  614.7  11.3  25.6  42.0   4.0   0.0   0.00
516  6  3   nov  tue  79.5    3.0  106.7   1.1  11.8  31.0   4.5   0.0   0.00

[517 rows x 13 columns]


max_rows
The number of rows returned is defined in Pandas option settings.

You can check your system's maximum rows with the pd.options.display.max_rows statement.

In [25]:
# Check the number of maximum returned rows:

import pandas as pd

print(pd.options.display.max_rows) 

60


In my system the number is 60, which means that if the DataFrame contains more than 60 rows, the print(df) statement will return only the headers and the first and last 5 rows.

You can change the maximum rows number with the same statement.

In [26]:
# Increase the maximum number of rows to display the entire data DataFrame:

import pandas as pd

pd.options.display.max_rows = 9999

print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [27]:
# Increase the maximum number of rows to display the entire forestfires DataFrame:

# Code Here
pd.options.display.max_rows = 9999

print(forestfires)

     X  Y month  day  FFMC    DMC     DC   ISI  temp     RH  wind  rain  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2   51.0   6.7   0.0   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0   33.0   0.9   0.0   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6   33.0   1.3   0.0   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3   97.0   4.0   0.2   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4   99.0   1.8   0.0   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2   29.0   5.4   0.0   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1   27.0   3.1   0.0   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0   86.0   2.2   0.0   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1   63.0   5.4   0.0   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8   40.0   4.0   0.0   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8   51.0   7.2   0.0   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3   38.0   4.0   0.0   
12   6  5   aug  fri  63.

**Pandas - Analyzing DataFrames**

**Viewing the Data**

One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

In [28]:
# Get a quick overview by printing the first 10 rows of the data DataFrame:

import pandas as pd

print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [29]:
# Get a quick overview by printing the first 10 rows of the forestfires DataFrame:

# Code Here
print(forestfires.head(10))

   X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain  area
0  7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51.0   6.7   0.0   0.0
1  7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33.0   0.9   0.0   0.0
2  7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33.0   1.3   0.0   0.0
3  8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97.0   4.0   0.2   0.0
4  8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99.0   1.8   0.0   0.0
5  8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29.0   5.4   0.0   0.0
6  8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27.0   3.1   0.0   0.0
7  8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86.0   2.2   0.0   0.0
8  8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63.0   5.4   0.0   0.0
9  7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40.0   4.0   0.0   0.0


In [30]:
# Print the first 5 rows of the data DataFrame:

import pandas as pd

print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


In [31]:
# Print the first 5 rows of the forestfires DataFrame:

# Code Here
print(forestfires.head())

   X  Y month  day  FFMC   DMC     DC  ISI  temp    RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51.0   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33.0   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33.0   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97.0   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99.0   1.8   0.0   0.0


There is also a tail() method for viewing the last rows of the DataFrame.

The tail() method returns the headers and a specified number of rows, starting from the bottom.

In [32]:
# Print the last 5 rows of the data DataFrame:

print(df.tail()) 

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


In [33]:
# Print the last 5 rows of the forestfires DataFrame:

# Code Here
print(forestfires.tail())

     X  Y month  day  FFMC    DMC     DC   ISI  temp    RH  wind  rain   area
512  4  3   aug  sun  81.6   56.7  665.6   1.9  27.8  32.0   2.7   0.0   6.44
513  2  4   aug  sun  81.6   56.7  665.6   1.9  21.9  71.0   5.8   0.0  54.29
514  7  4   aug  sun  81.6   56.7  665.6   1.9  21.2  70.0   6.7   0.0  11.16
515  1  4   aug  sat  94.4  146.0  614.7  11.3  25.6  42.0   4.0   0.0   0.00
516  6  3   nov  tue  79.5    3.0  106.7   1.1  11.8  31.0   4.5   0.0   0.00


**Info About the Data**

The DataFrames object has a method called info(), that gives you more information about the data set.

In [34]:
# Print information about the data:

print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


The result tells us there are 169 rows and 4 columns:

And the name of each column, with the data type:

\

**Null Values**

The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data, and you will learn more about that in the next chapters.

In [35]:
# Print information about the forestfires:

# Code Here
print(forestfires.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   X       517 non-null    int64  
 1   Y       517 non-null    int64  
 2   month   517 non-null    object 
 3   day     517 non-null    object 
 4   FFMC    515 non-null    float64
 5   DMC     505 non-null    float64
 6   DC      505 non-null    float64
 7   ISI     511 non-null    float64
 8   temp    502 non-null    float64
 9   RH      506 non-null    float64
 10  wind    513 non-null    float64
 11  rain    515 non-null    float64
 12  area    517 non-null    float64
dtypes: float64(9), int64(2), object(2)
memory usage: 52.6+ KB
None



\
**Pandas - Cleaning Data**

**Data Cleaning**

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates


**Pandas - Cleaning Empty Cells**

**Empty Cells**

Empty cells can potentially give you a wrong result when you analyze data.

**Remove Rows**

One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [36]:
# Return a new Data Frame with no empty cells:

import pandas as pd

df = pd.read_csv('data.csv')

new_df = df.dropna()

print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [37]:
# Return a forestfires new Data Frame with no empty cells:

# Code Here
forestfires = forestfires.dropna()
print(forestfires.info())

<class 'pandas.core.frame.DataFrame'>
Index: 471 entries, 0 to 516
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   X       471 non-null    int64  
 1   Y       471 non-null    int64  
 2   month   471 non-null    object 
 3   day     471 non-null    object 
 4   FFMC    471 non-null    float64
 5   DMC     471 non-null    float64
 6   DC      471 non-null    float64
 7   ISI     471 non-null    float64
 8   temp    471 non-null    float64
 9   RH      471 non-null    float64
 10  wind    471 non-null    float64
 11  rain    471 non-null    float64
 12  area    471 non-null    float64
dtypes: float64(9), int64(2), object(2)
memory usage: 51.5+ KB
None


Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

In [38]:
# Remove all rows with NULL values:

import pandas as pd

df = pd.read_csv('data.csv')

df.dropna(inplace = True)

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [39]:
# Remove all rows with NULL values (forestfires.csv):

import pandas as pd

forestfires = pd.read_csv('forestfires.csv')

# Code Here
forestfires.dropna(inplace=True)
print(forestfires.info())

<class 'pandas.core.frame.DataFrame'>
Index: 471 entries, 0 to 516
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   X       471 non-null    int64  
 1   Y       471 non-null    int64  
 2   month   471 non-null    object 
 3   day     471 non-null    object 
 4   FFMC    471 non-null    float64
 5   DMC     471 non-null    float64
 6   DC      471 non-null    float64
 7   ISI     471 non-null    float64
 8   temp    471 non-null    float64
 9   RH      471 non-null    float64
 10  wind    471 non-null    float64
 11  rain    471 non-null    float64
 12  area    471 non-null    float64
dtypes: float64(9), int64(2), object(2)
memory usage: 51.5+ KB
None


Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.

\

**Replace Empty Values**

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

In [40]:
# Replace NULL values with the number 130:

import pandas as pd

df = pd.read_csv('data.csv')

df.fillna(130, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     130.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     130.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [41]:
# Replace NULL values with the number 74 (forestfires.csv):

# Code Here
forestfires.fillna(value=74, inplace=True)

**Replace Only For Specified Columns**

The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

In [42]:
# Replace NULL values in the "Calories" columns with the number 130:

import pandas as pd

df = pd.read_csv('data.csv')

df["Calories"].fillna(130, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     130.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     130.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [43]:
# Replace NULL values in the (forestfires.csv) "temp" columns with the number 18:

# Code Here
forestfires["temp"].fillna(value=18, inplace=True)

**Replace Using Mean, Median, or Mode**

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

In [44]:
# Calculate the MEAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mean()

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse     Calories
0          60    110       130   409.100000
1          60    117       145   479.000000
2          60    103       135   340.000000
3          45    109       175   282.400000
4          45    117       148   406.000000
5          60    102       127   300.000000
6          60    110       136   374.000000
7          45    104       134   253.300000
8          30    109       133   195.100000
9          60     98       124   269.000000
10         60    103       147   329.300000
11         60    100       120   250.700000
12         60    106       128   345.300000
13         60    104       132   379.300000
14         60     98       123   275.000000
15         60     98       120   215.200000
16         60    100       120   300.000000
17         45     90       112   375.790244
18         60    103       123   323.000000
19         45     97       125   243.000000
20         60    108       131   364.200000
21         45    100       119   282.000000
22         60    130       101   300.000000
23         45    105       132   246.000000
24         60    102       126   334.500000
25         60    100       120   250.000000
26         60     92       118   241.000000
27         60    103       132   375.790244
28         60    100       132   280.000000
29         60    102       129   380.300000
30         60     92       115   243.000000
31         45     90       112   180.100000
32         60    101       124   299.000000
33         60     93       113   223.000000
34         60    107       136   361.000000
35         60    114       140   415.000000
36         60    102       127   300.000000
37         60    100       120   300.000000
38         60    100       120   300.000000
39         45    104       129   266.000000
40         45     90       112   180.100000
41         60     98       126   286.000000
42         60    100       122   329.400000
43         60    111       138   400.000000
44         60    111       131   397.000000
45         60     99       119   273.000000
46         60    109       153   387.600000
47         45    111       136   300.000000
48         45    108       129   298.000000
49         60    111       139   397.600000
50         60    107       136   380.200000
51         80    123       146   643.100000
52         60    106       130   263.000000
53         60    118       151   486.000000
54         30    136       175   238.000000
55         60    121       146   450.700000
56         60    118       121   413.000000
57         45    115       144   305.000000
58         20    153       172   226.400000
59         45    123       152   321.000000
60        210    108       160  1376.000000
61        160    110       137  1034.400000
62        160    109       135   853.000000
63         45    118       141   341.000000
64         20    110       130   131.400000
65        180     90       130   800.400000
66        150    105       135   873.400000
67        150    107       130   816.000000
68         20    106       136   110.400000
69        300    108       143  1500.200000
70        150     97       129  1115.000000
71         60    109       153   387.600000
72         90    100       127   700.000000
73        150     97       127   953.200000
74         45    114       146   304.000000
75         90     98       125   563.200000
76         45    105       134   251.000000
77         45    110       141   300.000000
78        120    100       130   500.400000
79        270    100       131  1729.000000
80         30    159       182   319.200000
81         45    149       169   344.000000
82         30    103       139   151.100000
83        120    100       130   500.000000
84         45    100       120   225.300000
85         30    151       170   300.000000
86         45    102       136   234.000000
87        120    100       157  1000.100000
88         45    129       103   242.000000
89         20     83       107    50.300000
9

In [45]:
# Calculate the MEAN, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
mean = forestfires["temp"].mean()
forestfires["temp"].fillna(value=mean, inplace=True)


Mean = the average value (the sum of all values divided by number of values).

In [46]:
# Calculate the MEDIAN, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].median()

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     318.6
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     318.6
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     318.6
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [47]:
# Calculate the MEDIAN, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
median = forestfires["temp"].median()
forestfires.fillna(value=median, inplace=True)

Median = the value in the middle, after you have sorted all values ascending.

In [48]:
# Calculate the MODE, and replace any empty values with it:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mode()[0]

df["Calories"].fillna(x, inplace = True)

df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132     300.0
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.0
37         60    100       120     300.0
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.0
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137     300.0
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [49]:
# Calculate the MODE, and replace any empty values with it (forestfires) "temp" columns:

# Code Here
firstMode = forestfires["temp"].mode()[0]
forestfires.fillna(value=firstMode, inplace=True)

Mode = the value that appears most frequently.

\
**Pandas - Cleaning Data of Wrong Format**

**Data of Wrong Format**

Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

**Convert Into a Correct Format**

In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Let's try to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

We will use 'data_Date.csv' dataset.

In [50]:
import pandas as pd

df = pd.read_csv('data_Date.csv')
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [2]:
# Convert to date:

import pandas as pd

df = pd.read_csv('data_Date.csv')

df['Date'] = pd.to_datetime(df['Date'])

print(df.to_string())

ValueError: time data "20201226" doesn't match format "'%Y/%m/%d'", at position 26. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

As you can see from the result, the date in row 26 was fixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row.

\

Removing Rows
The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the dropna() method.

In [ ]:
# Remove rows with a NULL value in the "Date" column:

df.dropna(subset=['Date'], inplace = True)

df

**Pandas - Fixing Wrong Data**

**Wrong Data**

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.

How can we fix wrong values, like the one for "Duration" in row 7?



**Replacing Values**

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

In [ ]:
# Set "Duration" = 45 in row 7:

df.loc[7, 'Duration'] = 45

df

In [ ]:
# Set "Duration" = 55 in row 17 and 20:

# Code Here
df.loc[17, "Duration"] = 55
df.loc[20, "Duration"] = 55
df

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

In [ ]:
# Loop through all values in the "Duration" column.

# If the value is higher than 120, set it to 120:

# Code Here
for index, row in df.iterrows():
    if row["Duration"] > 120:
        row["Duration"] = 120

**Removing Rows**

Another way of handling wrong data is to remove the rows that contains wrong data.

This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

In [ ]:
# Delete rows where "Duration" is higher than 120:

# Code Here
for index, row in df.iterrows():
    if row["Duration"] > 120:
        df.drop(index)

**Pandas - Removing Duplicates**

**Discovering Duplicates**

Duplicate rows are rows that have been registered more than one time.

\

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

In [ ]:
# Returns True for every row that is a duplicate, othwerwise False:

print(df.duplicated())

In [ ]:
# Returns True for every row that is a duplicate, othwerwise False (forestfires.csv):

# Code Here
print(forestfires.duplicated())

**Removing Duplicates**

To remove duplicates, use the drop_duplicates() method.

In [ ]:
# Remove all duplicates:

df.drop_duplicates(inplace = True)

df

In [ ]:
# Remove all duplicates (forestfires.csv):

# Code Here
forestfires.drop_duplicates(inplace=True)
print(forestfires.duplicated())

**Pandas - Data Correlations**

Finding Relationships
A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

The examples uses a CSV file called: 'data.csv'.

In [53]:
# Show the relationship between the columns:

import pandas as pd

df = pd.read_csv('data.csv')

df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.155408  0.009403  0.922717
Pulse    -0.155408  1.000000  0.786535  0.025121
Maxpulse  0.009403  0.786535  1.000000  0.203813
Calories  0.922717  0.025121  0.203813  1.000000

Note: The corr() method ignores "not numeric" columns.

**Result Explained**

The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

\

**What is a good correlation?**

 It depends on the use, but I think it is safe to say you have to have at least 0.6 (or -0.6) to call it a good correlation.

**Perfect Correlation:**

We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

**Good Correlation:**

"Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

**Bad Correlation:**

"Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.

In [70]:
# Show the relationship between the columns (forestfires.csv):

# Code Here
forestfires.corr()

#otherwise causes errors if month and day columns present hence we drop it first
#forestfires.drop("month", axis=1, inplace=True)
#forestfires.drop("day", axis=1, inplace=True)

X         Y      FFMC       DMC        DC       ISI      temp  \
X     1.000000  0.527241 -0.013076 -0.040276 -0.086498  0.003541 -0.056289   
Y     0.527241  1.000000 -0.037023  0.018252 -0.085425 -0.018162 -0.021473   
FFMC -0.013076 -0.037023  1.000000  0.390568  0.345502  0.522484  0.431088   
DMC  -0.040276  0.018252  0.390568  1.000000  0.686080  0.310055  0.471407   
DC   -0.086498 -0.085425  0.345502  0.686080  1.000000  0.231568  0.506275   
ISI   0.003541 -0.018162  0.522484  0.310055  0.231568  1.000000  0.387261   
temp -0.056289 -0.021473  0.431088  0.471407  0.506275  0.387261  1.000000   
RH    0.089642  0.065051 -0.301125  0.060116 -0.066979 -0.141793 -0.553375   
wind  0.013198 -0.037228 -0.010280 -0.096468 -0.189836  0.118715 -0.217425   
rain  0.066988  0.032179  0.058032  0.075377  0.037542  0.069029  0.070949   
area  0.054653  0.031042  0.040729  0.069484  0.051400  0.005769  0.101916   

            RH      wind      rain      area  
X     0.089642  0.013198  0.066988  0.054653  
Y     0.065051 -0.037228  0.032179  0.031042  
FFMC -0.301125 -0.010280  0.058032  0.040729  
DMC   0.060116 -0.096468  0.075377  0.069484  
DC   -0.066979 -0.189836  0.037542  0.051400  
ISI  -0.141793  0.118715  0.069029  0.005769  
temp -0.553375 -0.217425  0.070949  0.101916  
RH    1.000000  0.088469  0.102996 -0.078480  
wind  0.088469  1.000000  0.064726  0.005454  
rain  0.102996  0.064726  1.000000 -0.008226  
area -0.078480  0.005454 -0.008226  1.000000